api/src/prescription/learner-management/domain/models/CommonOrganizationLearnerSet.js

```diff
@@ -0,0 +1,88 @@
+import { EntityValidationRulesError } from '../../../../shared/domain/errors.js';
+import { validateCommonOrganizationLearner } from '../validators/common-organization-learner-validator.js';
+
+class ImportOrganizationLearnerSet {
+  #learners;
+  #unicityKeys;
+  #hasValidationFormats;
+  #hasUnicityRules;
+
+  constructor(validationRules) {
+    this.#learners = [];
+    this.#unicityKeys = [];
+    this.validationRules = validationRules;
+    this.#hasUnicityRules = !!validationRules?.unicity;
+    this.#hasValidationFormats = !!validationRules?.formats;
+  }
+
+  addLearner(learnerAttributes) {
+    this.#validateRules(learnerAttributes);
+    this.#learners.push(new CommonOrganizationLearner(learnerAttributes));
+  }
+
+  #validateRules(learnerAttributes) {
+    const errors = [];
+
+    if (this.#hasUnicityRules) {
+      const unicityError = this.#checkUnicityRule(learnerAttributes);
+
+      if (unicityError) {
+        errors.push(unicityError);
+      }
+    }
+
+    if (this.#hasValidationFormats) {
+      const validationErrors = this.#checkValidations(learnerAttributes);
+
+      if (validationErrors) {
+        errors.push(...validationErrors);
+      }
+    }
+
+    if (errors.length > 0) {
+      throw errors;
+    }
+  }
+
+  #checkUnicityRule(learnerAttributes) {
+    const unicityKeys = [];
+    this.validationRules.unicity.forEach((rule) => {
+      unicityKeys.push(learnerAttributes.attributes[rule]);
+    });
+    const unicityEntity = unicityKeys.join('-');
+    if (!this.#unicityKeys.includes(unicityEntity)) {
+      this.#unicityKeys.push(unicityEntity);
+      return null;
+    } else {
+      return EntityValidationRulesError.unicityError({
+        key: this.validationRules.unicity.join('-'),
+      });
+    }
+  }
```

Comment:
**thought** : J'ai vraiment eu beaucoup de difficulté à comprendre ce que faisait cette méthode. Si on doit y revenir plus tard j'ai le sentiment que ça risque d'être compliqué 🤔

Reply:
J'ai le sentiment qu'on pourrait faire quelque chose de ce type là : 
```js
const checkNotUniq(learnerToCheck) => {
  return (learner) => this.validationRules.unicity.some(rule => learner[rule] === learnerToCheck[rule])
}

if (this.#learners.some(checkNotUniq(learnerAttributes))) {
  return EntityValidationRulesError.unicityError({
    key: this.validationRules.unicity.join('-'),
  });
}
```

C'est peut être moins performant mais ça me semble tôt pour chercher la performance 🤔 Qu'en pensez-vous ?

Reply:
On était reparti sur la même manière dont était vérifié l'unicité du SUP ( excepté la concatenation des clés d'unicité 🆕 ).

Peut être que le nommage est pas terrible.

Reply:
Hello 👋 
Nous allons repartir sur la méthode précédente après avoir testé la méthode proposée.

Dans cette dernière, pour effectuer la comparaison, nous utilisons le `this.#learners` dans lequel sont stockés uniquement les learners respectant toutes les règles de validation. Ce qui exclu les learners qui ne les respectent pas.
De ce fait si on a un `learner B` ne répondant pas à la règle d'unicité par rapport à un `learner A`, ne respectant pas de son côté la règle de format de donnée, l'erreur sur la règle d'unicité ne sera pas remontée car le `learner A` ne sera pas dans `this.#learners` permettant la comparaison.

Tricky à expliquer 🤯

Reply:
Le refacto de ce bout de code est ajouté dans le fichier dette technique